api/python/t4/search_util.py

```diff
@@ -0,0 +1,86 @@
+"""
+search_util.py
+
+Contains search-related glue code
+"""
+import json
+from urllib.parse import urlparse
+
+from aws_requests_auth.boto_utils import BotoAWSRequestsAuth
+from elasticsearch import Elasticsearch, RequestsHttpConnection
+
+ES_INDEX = 'drive'
+
+def _create_es(search_endpoint, aws_region):
+    """
+    search_endpoint: url for search endpoint
+    aws_region: name of aws region endpoint is hosted in
+    """
+    es_url = urlparse(search_endpoint)
+
+    auth = BotoAWSRequestsAuth(aws_host=es_url.hostname,
+                               aws_region=aws_region,
+                               aws_service='es')
+    port = es_url.port or (443 if es_url.scheme == 'https' else 80)
+
+    es_client = Elasticsearch(
+        hosts=[{'host': es_url.hostname, 'port': port}],
+        http_auth=auth,
+        use_ssl=True,
+        verify_certs=True,
+        connection_class=RequestsHttpConnection
+    )
+
+    return es_client
+
+def search(query, search_endpoint, aws_region='us-east-1'):
+    """
+    Searches your bucket. query can contain plaintext, and can also contain clauses
+    like $key:"$value" that search for exact matches on specific keys.
+
+    Returns either the request object (in case of an error)
+            or a list of objects with the following keys:
+        key: key of the object
+        version_id: version_id of object version
+        operation: Create or Delete
+        meta: metadata attached to object
+        size: size of object in bytes
+        text: indexed text of object
+        source: source document for object (what is actually stored in ElasticSeach)
+        time: timestamp for operation
+    """
+    es_client = _create_es(search_endpoint, aws_region)
+
+    payload = {'query': {'query_string': {
+        'default_field': 'content',
+        'query': query,
+        'quote_analyzer': 'keyword',
+        }}}
+
+    raw_response = es_client.search(index=ES_INDEX, body=payload)
+
+    try:
+        results = []
+        for result in raw_response['hits']['hits']:
+            key = result['_source']['key']
+            vid = result['_source']['version_id']
+            operation = result['_source']['type']
+            meta = json.dumps(result['_source']['user_meta'])
```

Comment:
Is this field guaranteed to be set?

Reply:
yes it will be `{}` if the indexer doesn't see any user meta